service/geopmdpy_test/TestSession.py

```diff
@@ -108,33 +98,40 @@ def test_format_signals(self):
             pfs.assert_has_calls(calls)
 
     def test_run_read(self):
-        reqs_return_value = [1, 2, 3]
-        requests = mock.MagicMock(return_value=reqs_return_value)
-
-        format_return_value = [4, 5, 6]
-        requests.get_formats.return_value = format_return_value
-
         duration = 2
         period = 1
         num_period = int(duration / period)
         out_stream = mock.MagicMock()
 
-        with mock.patch('geopmdpy.runtime.TimedLoop', return_value=[0, 1]) as rtl, \
-             mock.patch('geopmdpy.session.Session.read_signals',
-                        return_value=reqs_return_value) as srs, \
-             mock.patch('geopmdpy.session.Session.format_signals',
-                        return_value=format_return_value) as sfs:
-            self._session.run_read(requests, duration, period, out_stream)
-
-            self._geopm_proxy.PlatformOpenSession.assert_called_once_with()
-            rtl.assert_called_once_with(period, num_period)
-            srs.assert_has_calls([mock.call(requests)] * num_period)
-            sfs.assert_has_calls([mock.call(reqs_return_value, format_return_value)] * num_period)
-            out_stream.write.assert_has_calls([mock.call(format_return_value)] * num_period)
-            self._geopm_proxy.PlatformCloseSession.assert_called_once_with()
+        user_requests = [('power', 0, 0), ('SERVICE::energy', 1, 1), ('frequency', 2, 2)]
+        pio_requests = [('SERVICE::power', 0, 0), ('SERVICE::energy', 1, 1), ('SERVICE::frequency', 2, 2)]
+
+        mock_requests = mock.MagicMock()
+        mock_requests.__iter__.return_value = user_requests
+        format_return_value = "1.234, 2.345, 3.456"
+        mock_requests.get_formats.return_value = format_return_value
+        signal_handle = range(3)
+        signal_expect = num_period * [1.234, 2.345, 3.456]
+
+        with mock.patch('geopmdpy.runtime.TimedLoop', return_value=[0, 1]) as mock_timed_loop, \
+             mock.patch('geopmdpy.pio.push_signal', side_effect=signal_handle) as mock_push_signal, \
+             mock.patch('geopmdpy.pio.read_batch') as mock_read_batch, \
+             mock.patch('geopmdpy.pio.sample', side_effect=signal_expect) as mock_sample, \
+             mock.patch('geopmdpy.session.Session.format_signals', return_value=format_return_value):
+            # Call tested method
+            self._session.run_read(mock_requests, duration, period, out_stream)
+            # Check mocks calls
+            calls = [mock.call(*req) for req in pio_requests]
+            mock_push_signal.assert_has_calls(calls)
+            calls = num_period * [mock.call()]
+            mock_read_batch.assert_has_calls(calls)
+            calls = num_period * [mock.call(idx) for idx in signal_handle]
```

Comment:
This looks suspicious, since signal_handle is a generator with 3 values which should be exhausted by mock push_signal. Can you verify that it doesn't evaluate to `calls = num_period * []`? If that does happen, I think the next assert would be a no-op.

Reply:
So maybe just change line 113 to be

```
signal_handle = list(range(3))
```

So this is unambiguous?